homeassistant/components/notion/binary_sensor.py

```diff
@@ -0,0 +1,92 @@
+"""Support for Notion binary sensors."""
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+
+from . import (
+    ATTR_BRIDGE_MODE, ATTR_BRIDGE_NAME, BINARY_SENSOR_TYPES, SENSOR_BATTERY,
+    SENSOR_DOOR, SENSOR_GARAGE_DOOR, SENSOR_LEAK, SENSOR_MISSING, SENSOR_SAFE,
+    SENSOR_SLIDING, SENSOR_SMOKE_CO, SENSOR_WINDOW_HINGED_HORIZONTAL,
+    SENSOR_WINDOW_HINGED_VERTICAL, NotionEntity)
+
+from .const import DATA_CLIENT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Notion sensors based on a config entry."""
+    notion = hass.data[DOMAIN][DATA_CLIENT][entry.entry_id]
+
+    sensor_list = []
+    for task in notion.tasks:
+        if task['task_type'] not in BINARY_SENSOR_TYPES:
+            continue
+
+        name, device_class = BINARY_SENSOR_TYPES[task['task_type']]
+        sensor = next(
+            (s for s in notion.sensors if s['id'] == task['sensor_id'])
+        )
+        bridge = next(
+            (b for b in notion.bridges if b['id'] == sensor['bridge']['id'])
+        )
+        system = next(
+            (s for s in notion.systems if s['id'] == sensor['system_id'])
+        )
+
+        sensor_list.append(
+            NotionBinarySensor(
+                notion, task, sensor, bridge, system, name, device_class))
+
+    async_add_entities(sensor_list, True)
+
+
+class NotionBinarySensor(NotionEntity, BinarySensorDevice):
+    """Define a Notion sensor."""
+
+    @property
+    def is_on(self):
+        """Return whether the sensor is on or off."""
+        if self._task['task_type'] == SENSOR_BATTERY:
+            return self._state != 'battery_good'
+        if self._task['task_type'] in (
+                SENSOR_DOOR, SENSOR_GARAGE_DOOR, SENSOR_SAFE, SENSOR_SLIDING,
+                SENSOR_WINDOW_HINGED_HORIZONTAL,
+                SENSOR_WINDOW_HINGED_VERTICAL):
+            return self._state != 'closed'
+        if self._task['task_type'] == SENSOR_LEAK:
+            return self._state != 'no_leak'
+        if self._task['task_type'] == SENSOR_MISSING:
+            return self._state == 'not_missing'
+        if self._task['task_type'] == SENSOR_SMOKE_CO:
+            return self._state != 'no_alarm'
+
+    async def async_update(self):
+        """Fetch new state data for the sensor."""
+        try:
+            new_task_data = next(
+                (t for t in self._notion.tasks if t['id'] == self._task['id'])
```

Comment:
I see a lot of iteration through tasks, any reason you're storing it as a list instead of converting it to a dict when the data is loaded for faster lookups ? `{t['id']: t for t in self._notion.tasks}`

Reply:
Outstanding point. Let me know if https://github.com/home-assistant/home-assistant/pull/24634/commits/96b2fdc49286dce894aace1e73ec9bc05f379443 looks good?